util/fpga/cw_spiflash.py

```diff
@@ -0,0 +1,273 @@
+      is_second_frame: Indicates if this is the second frame.
+      expected_ack: Expected acknowledgement value for a frame.
+    """
+    HASH_FUNCTION = hashlib.sha256
+    DIGEST_SIZE = HASH_FUNCTION().digest_size
+    FRAME_NUMBER_SIZE = 4
+    FLASH_OFFSET_SIZE = 4
+    FRAME_SIZE = 2048
+    PAYLOAD_SIZE = FRAME_SIZE - DIGEST_SIZE - FRAME_NUMBER_SIZE - FLASH_OFFSET_SIZE
+
+    def __init__(self, frame_number, flash_offset, payload):
+        """Inits a _SpiFlashFrame.
+
+        Also pads frame payload to ``PAYLOAD_SIZE`` bytes if needed.
+
+        Args:
+          frame_numer: Frame number.
+          flash_offset: Flash offset.
+          payload: Frame payload.
+        """
+        self._frame_number = _uint_to_le(frame_number, self.FRAME_NUMBER_SIZE)
+        self._flash_offset = _uint_to_le(flash_offset, self.FLASH_OFFSET_SIZE)
+        self._payload = payload
+        padding = self.PAYLOAD_SIZE - len(self._payload)
+        if padding > 0:
+            self._payload = b''.join([self._payload, b'\xFF' * padding])
+        self._digest = self.HASH_FUNCTION(b''.join([
+            self._frame_number,
+            self._flash_offset,
+            self._payload,
+        ])).digest()
+
+    def __bytes__(self):
+        return b''.join([
+            self._digest,
+            self._frame_number,
+            self._flash_offset,
+            self._payload,
+        ])
+
+    def __repr__(self):
+        return f'frame 0x{_le_to_uint(self._frame_number):08X} @ \
+            0x{_le_to_uint(self._flash_offset):08X}'
+
+    @property
+    def is_second_frame(self):
+        """Indicates if this is the second frame."""
+        return _le_to_uint(self._frame_number) == 1
+
+    @property
+    def expected_ack(self):
+        """Expected acknowledgement value for the frame."""
+        return self.HASH_FUNCTION(bytes(self)).digest()
+
+
+class _Bootstrap:
+    """Handles low-level details during programming OpenTitan using SAM3X/U on CW310/305.
+
+    Initializes pins, resets OpenTitan, transmits frames, and receives acknowledgements.
+
+    To use:
+    >>> with _Bootstrap(fpga) as bootstrap:
+    >>>   ...
+    >>>   ack = bootstrap.transfer(frame)
+    >>>   ...
+    """
+    # Default pin mapping for CW305 board.
+    _PIN_SCK = 'USB_A13'
+    _PIN_SDI = 'USB_A14'
+    _PIN_SDO = 'USB_A15'
+    _PIN_CS = 'USB_A16'
+    _PIN_TRST = 'USB_A17'
+    _PIN_SRST = 'USB_A18'
+    _PIN_JTAG_SPI = 'USB_A19'
+    _PIN_BOOTSTRAP = 'USB_A20'
+    # Delays below are in seconds.
+    _BOOTSTRAP_DELAY = 0.1
+    _SECOND_FRAME_DELAY = 0.2
+    _INTER_FRAME_DELAY = 0.02
+
+    def __init__(self, fpga, board="CW305"):
+        """Inits a _Bootstrap with a CW310/305.
+
+        Args:
+          fpga: CW310/305 to be programmed, a ``cw.capture.targets.CW310/305`` instance.
+          board: can be ``CW310`` or ``CW305`` to distinguish the different board types.
+        """
+
+        # Change pin mapping to CW310 board.
+        if board == "CW310":
```

Comment:
I think this should be a different mapping defined above (maybe using namedtuples for each board), and check that board is either CW310 or CW305.